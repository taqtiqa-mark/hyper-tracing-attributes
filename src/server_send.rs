pub use analyze::{analyze, Model};
pub use codegen::codegen;
pub use lower::{lower, Ir};
pub use parsed::{parse, parse_args, Ast};

pub mod analyze;
pub mod codegen;
pub mod lower;
pub mod parsed;
