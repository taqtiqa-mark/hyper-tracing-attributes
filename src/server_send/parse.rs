use proc_macro2::TokenStream;
use proc_macro_error::{abort, abort_call_site};
use syn::{Expr, Item, ItemFn};

pub type Ast = ItemFn;

// Attribute arguments that are assignment expressions.
pub struct Args {
    vars: std::collections::HashSet<syn::ExprAssign>
}

// Parse the attribute data once we know it contains ExprAssign
impl syn::parse::Parse for Args {
    fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
        // parses a=1,b="d",c=f(), where a,b and c are ExprAssign
        let vars = syn::punctuated::Punctuated::<syn::ExprAssign, syn::Token![,]>::parse_terminated(input)?;
        Ok(Args {
            vars: vars.into_iter().collect(),
        })
    }
}

// Using newly created struct Args, parse assignment arguments then count.
// This utility/no-op function is required by the fact
// `syn::parse_macro_input!` must be called in a function that returns
// proc_macro::TokenStream.
pub fn parse_args(metadata: proc_macro::TokenStream) -> proc_macro::TokenStream {
    const ERROR: &str = "this attribute takes one, two or three arguments";
    const HELP: &str = "use `#[server_send(level=trace, name=\"Name\", skip=[a,b])]`";

    eprintln!("{}", metadata);
    let expr = syn::parse2::<syn::ExprAssign>(metadata.clone().into()).expect("ExprAssign parsing");
    let args = syn::parse_macro_input!(metadata as Args);

    if args.vars.len() > 3 {
        // ../tests/ui/server_send/error/has-too-many-arguments.rs
        abort!(expr, ERROR; help = HELP)
    }
    proc_macro::TokenStream::from(quote::quote!{fn dummy(){}})
}

pub fn parse(metadata: TokenStream, item: TokenStream) -> Ast {
    const ERROR: &str = "this attribute takes one, two or three arguments";
    const HELP: &str = "use `#[server_send(level=trace, name=\"Name\", skip=[a,b])]`";

    if metadata.is_empty() {
        // ../tests/ui/server_send/error/has-no-arguments.rs
        abort_call_site!(ERROR; help = HELP)
    }

    let expr = syn::parse2::<Expr>(metadata.clone()).expect("Expr parsing");
    match expr {
        Expr::Array(expr) => {
            // ../tests/ui/server_send/error/has-expression-arguments.rs
            abort!(expr, ERROR; help = HELP)
        }
        Expr::Assign(_expr) => {
            parse_args(metadata.into());
        }
        _ => {
            // ../tests/ui/server_send/error/has-other-arguments.rs
            abort!(expr, ERROR; help = HELP)
        }
    }


    match syn::parse2::<Item>(item) {
        Ok(Item::Fn(item)) => item,
        Ok(item) => {
            // ../tests/ui/server_send/error/item-is-not-a-function.rs
            abort!(
                item,
                "item is not a function";
                help = "`#[server_send(...)]` can only be used on functions"
            )
        }
        Err(_) => unreachable!(), // ?
    }
}

#[cfg(test)]
mod tests {
    use quote::quote;

    use super::*;

    #[test]
    fn valid_field_syntax() {
        parse(
            quote!(),
            quote!(
                #[inline]
                #[trace_field(x = 0)]
                fn even_to_odd(x: u32) -> u32 {
                    x + 1
                }
            ),
        );
    }
    #[test]
    fn valid_instrument_syntax() {
        parse(
            quote!(),
            quote!(
                #[inline]
                #[server_send(level = trace, name = "Neat", skip = [x])]
                fn even_to_odd(x: u32) -> u32 {
                    x + 1
                }
            ),
        );
    }
}
