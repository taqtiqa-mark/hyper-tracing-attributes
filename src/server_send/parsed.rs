use proc_macro2::TokenStream;
use proc_macro_error::{abort, abort_call_site};
use syn::{Expr, Item, ItemFn};
use darling::{FromMeta};

pub type Ast = ItemFn;

/// Describes the data that directly corresponds to the attributes of the
/// `server_send`-proc-macro.
/// This is the parsing target for the
/// crate "darling" which works together with `syn`.
/// FromMeta implements "from_list"-factory-method (`darling` crate).
#[derive(Debug, darling::FromMeta)]
struct TaskAttributes {
    #[darling(default)]
    level: Option<String>,
    #[darling(default)]
    name: Option<String>,
    #[darling(default)]
    skip: Option<String>,
    #[darling(default)]
    fields: Option<String>,
}

// #[derive(Debug, darling::FromMeta)]
// struct InstrumentAttributes {
//     #[darling(default)]
//     level: Option<syn::Expr>,
//     #[darling(default)]
//     name: Option<syn::Expr>,
//     #[darling(default)]
//     skip: Option<syn::Expr>,
//     #[darling(default)]
//     fields: Option<syn::Expr>,
// }

impl TaskAttributes {
    /// Maps property `skip` of `TaskAttributes` from a
    /// comma-separated string to a vector of strings.
    fn skip_params(&self) -> Vec<String> {
        self.skip
            .as_ref()
            .map(|s| s.clone())
            .map(|s| {
                s.split(",")
                    .into_iter()
                    .map(|s| s.to_owned())
                    .collect::<Vec<String>>()
            })
            .unwrap_or(vec![])
            .into_iter()
            .map(|s| s.trim().to_owned())
            .collect::<Vec<String>>()
    }

    /// Maps property `read` of `TaskAttributes` from a
    /// comma-separated string to a vector of strings.
    fn fields_params(&self) -> Vec<String> {
        self.fields
            .as_ref()
            .map(|s| s.clone())
            .map(|s| {
                s.split(",")
                    .into_iter()
                    .map(|s| s.to_owned())
                    .collect::<Vec<String>>()
            })
            .unwrap_or(vec![])
            .into_iter()
            .map(|s| s.trim().to_owned())
            .collect::<Vec<String>>()
    }
}

// /// Invoked like this:
// /// - `#[task(write = "data1, data2")]`
// /// - `#[task(write = "data1", read = "data2")]`
// /// - `#[task(read = "data2")]`
// /// If a parameter is write it is automatically also read.
// #[proc_macro_attribute]
// pub fn task(args: TokenStream, item: TokenStream) -> TokenStream {
//     // we do nothing here; just a marker for the super macro
//     let attr_args = parse_macro_input!(args as AttributeArgs);
//     let attrs: TaskAttributes =
//         TaskAttributes::from_list(&attr_args).expect("#[task] could not be parsed!");
//     println!("read_params: {:#?}", attrs.read_params());
//     println!("write_params: {:#?}", attrs.write_params());
//     item
// }

// Attribute arguments that are assignment expressions.
#[derive(Debug)]
pub struct Args {
    vars: Vec<syn::Expr>
}

// Parse the attribute data once we know it contains ExprAssign
impl syn::parse::Parse for Args {
    fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
        // parses a=1,b="d",c=f(), where a,b and c are ExprAssign
        let vars = syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated(input)?;
        Ok(Args {
            vars: vars.into_iter().collect(),
        })
    }
}
// Parse attribute arguments then count.
// This utility/no-op function is required by the fact
// `syn::parse_macro_input!` must be called in a function that returns
// proc_macro::TokenStream.
pub fn parse_args(metadata: proc_macro::TokenStream) -> proc_macro::TokenStream {
    const ERROR: &str = "this attribute takes from one to four arguments";
    const HELP: &str = "use `#[server_send(level=trace, name=\"Name\", skip=\"a, b\")]`";

    eprintln!("{}", metadata);

    if metadata.is_empty() {
        // ../tests/ui/server_send/error/has-no-arguments.rs
        abort_call_site!(ERROR; help = HELP)
    }

    let md = metadata.clone();
    let args = syn::parse_macro_input!(md as Args);

    eprintln!("{}",args.vars.len());

    // if args.vars.len() > 3 {
    //     // ../tests/ui/server_send/error/has-too-many-arguments.rs
    //     abort_call_site!(ERROR; help = HELP)
    // }

    eprintln!("{:?}", args);

    for expr in &args.vars {
        match expr {
            Expr::Array(expr) => {
                // ../tests/ui/server_send/error/has-expression-arguments.rs
                abort!(expr, ERROR; help = HELP)
            }
            Expr::Assign(_expr) => { }
            _ => {
                // ../tests/ui/server_send/error/has-other-arguments.rs
                abort!(expr, ERROR; help = HELP)
            }
        }
    }
    // let expr: syn::Expr = syn::parse2(metadata.clone().into()).expect("ExprAssign parsing");

    // eprintln!("{:?}", expr);

    // let attr_args = syn::parse_macro_input!(metadata as syn::AttributeArgs);

    // eprintln!("{:?}", attr_args);

    // // let attr_args: syn::NestedMeta = args.vars.into();
    // let attrs: TaskAttributes =
    //     TaskAttributes::from_list(&attr_args).expect("#[server_send(...)] could not be parsed!");

    // eprintln!("{:?}", attrs);

    // eprintln!("skip_params: {:#?}", attrs.skip_params());
    // eprintln!("fields_params: {:#?}", attrs.fields_params());

    // if attr_args.len() > 4 {
    //     // ../tests/ui/server_send/error/has-too-many-arguments.rs
    //     abort_call_site!(ERROR; help = HELP)
    // }

    proc_macro::TokenStream::from(quote::quote! {fn dummy(){}})
}

pub fn parse(metadata: TokenStream, item: TokenStream) -> Ast {

    parse_args(metadata.clone().into());

    // let expr = syn::parse::<Expr>(metadata.clone().into()).expect("Expr parsing");
    // match expr {
    //     Expr::Array(expr) => {
    //         // ../tests/ui/server_send/error/has-expression-arguments.rs
    //         abort!(expr, ERROR; help = HELP)
    //     }
    //     Expr::Assign(_expr) => {
    //         parse_args(metadata.into());
    //     }
    //     _ => {
    //         // ../tests/ui/server_send/error/has-other-arguments.rs
    //         abort!(expr, ERROR; help = HELP)
    //     }
    // }

    match syn::parse2::<Item>(item) {
        Ok(Item::Fn(item)) => item,
        Ok(item) => {
            // ../tests/ui/server_send/error/item-is-not-a-function.rs
            abort!(
                item,
                "item is not a function";
                help = "`#[server_send(...)]` can only be used on functions"
            )
        }
        Err(_) => unreachable!(), // ?
    }
}

// #[cfg(test)]
// mod tests {
//     use quote::quote;
//     use proc_macro_error::proc_macro_error;

//     use super::*;

    // #[test]
    // // #[proc_macro_attribute]
    // // #[proc_macro_error(allow_not_macro)]
    // fn valid_minimal_syntax() {
    //     parse_args(
    //         quote!(
    //             #[inline]
    //             #[server_send(level = info)]
    //             fn even_to_odd(x: u32) -> u32 {
    //                 x + 1
    //             }
    //         ).into()
    //     );
    //     //proc_macro::TokenStream::from(quote::quote!{fn dummy(){}})
    // }

//     #[test]
//     #[proc_macro_attribute]
//     //#[proc_macro_error(allow_not_macro)]
//     fn valid_instrument_syntax() {
//         parse_args(
//             quote!(
//                 #[inline]
//                 #[server_send(level = trace, name = "Neat", skip = "x,y")]
//                 fn some_addition(x: u32, y: u32) -> u32 {
//                     x + y
//                 }
//             ).into()
//         );
//         //proc_macro::TokenStream::from(quote::quote!{fn dummy(){}})
//     }
// }

// Args { vars: {ExprAssign { attrs: [], left: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident { ident: "target", span: #0 bytes(94..100) }, arguments: None }] } }), eq_token: Eq, right: Lit(ExprLit { attrs: [], lit: Str(LitStr { token: "bulls-eye" }) }) }, ExprAssign { attrs: [], left: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident { ident: "level", span: #0 bytes(58..63) }, arguments: None }] } }), eq_token: Eq, right: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident { ident: "trace", span: #0bytes(64..69) }, arguments: None }] } }) }, ExprAssign { attrs: [], left: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident { ident: "name", span: #0 bytes(71..75) }, arguments: None }] } }), eq_token: Eq, right: Lit(ExprLit { attrs: [], lit: Str(LitStr { token: "B" }) }) }, ExprAssign { attrs: [], left: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident { ident: "skip", span: #0 bytes(81..85) }, arguments: None }] } }), eq_token: Eq, right: Array(ExprArray { attrs: [], bracket_token: Bracket, elems: [Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident { ident: "self", span: #0 bytes(87..91) }, arguments: None }] } })] }) }} }