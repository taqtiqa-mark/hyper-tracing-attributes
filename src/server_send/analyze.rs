use proc_macro_error::{abort, abort_call_site};
use syn::{
    parenthesized,
    parse::{Parse, ParseStream},
    spanned::Spanned,
    ItemFn,
};

use crate::server_send::Ast;

pub fn analyze(ast: Ast) -> Model {
    let mut trace_fields = vec![];

    let mut item = ast;
    let attrs = &mut item.attrs;

    eprintln!("{:?}", attrs);

    for index in (0..attrs.len()).rev() {
        if let Some(ident) = attrs[index].path.get_ident() {
            if ident.to_string().as_str() == "trace_field" {
                let attr = attrs.remove(index);
                let span = attr.tokens.span();

                if let Ok(arg) = syn::parse2::<AttributeArgument>(attr.tokens) {
                    trace_fields.push(arg.expr);
                } else {
                    // ../tests/ui/server_send/error/trace_field-is-not-an-assignment-expression.rs
                    abort!(
                        span,
                        "expected an assignment expression as argument";
                        help = "example: `#[trace_field(b = tracing::field::Empty)]`")
                }
            }
        }
    }

    // if trace_fields.is_empty() {
    //     // ../tests/ui/server_send/error/zero-trace-feilds.rs
    //     abort_call_site!(
    //         "no trace fields were specified";
    //         help = "add a `#[trace_field]`"
    //     )
    // }

    Model { trace_fields, item }
}

struct AttributeArgument {
    expr: syn::ExprAssign,
}

impl Parse for AttributeArgument {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let content;
        let _parenthesis = parenthesized!(content in input);

        Ok(AttributeArgument {
            expr: content.parse()?,
        })
    }
}

pub struct Model {
    pub trace_fields: Vec<syn::ExprAssign>,
    pub item: ItemFn,
}

#[cfg(test)]
mod tests {
    use syn::{parse_quote, Attribute};

    use super::*;

    #[test]
    fn can_extract_trace_field() {
        let model = analyze(parse_quote!(
            #[trace_field(x=0)]
            fn f(x: bool) {}
        ));

        let expected: &[syn::ExprAssign] = &[parse_quote!(x=0)];
        assert_eq!(expected, model.trace_fields);

        assert!(model.item.attrs.is_empty());
    }

    // Ensure attributes unrelated to this proc-macro are not removed or reordered
    #[test]
    fn non_dsl_attributes_are_preserved() {
        let model = analyze(parse_quote!(
            #[a]
            #[trace_field(x=1)]
            #[b]
            fn f(x: bool) {}
        ));

        let expected: &[Attribute] = &[parse_quote!(#[a]), parse_quote!(#[b])];
        assert_eq!(expected, model.item.attrs);
    }
}
