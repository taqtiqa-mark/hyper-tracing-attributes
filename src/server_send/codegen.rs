use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::{parse_quote, ItemFn, Stmt};

use crate::server_send::{lower::Field, Ir};

pub type Rust = TokenStream;

pub fn codegen(ir: Ir) -> Rust {
    let Ir { fields, item } = ir;

    let ItemFn {
        attrs,
        vis,
        sig,
        block,
    } = item;

    // https://github.com/rust-lang/rust/issues/17022
    quote!(
        #(#attrs)*
        #(#fields)*
        #vis #sig {
            #block
        }
    )
}

impl ToTokens for Field {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Field { expr, message } = self;
        let stmt: Stmt = parse_quote!(instrument(#expr, #message););
        stmt.to_tokens(tokens);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn output_is_function_item() {
        let ir = Ir {
            fields: vec![Field {
                expr: parse_quote!(x=0),
                message: "message".to_string(),
            }],
            item: parse_quote!(
                fn f() {}
            ),
        };
        let rust = codegen(ir);

        eprintln!("{}", rust);

        assert!(syn::parse2::<ItemFn>(rust).is_ok());
    }
}
