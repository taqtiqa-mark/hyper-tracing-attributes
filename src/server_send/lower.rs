use quote::quote;
use syn::{ItemFn};

use crate::server_send::Model;

pub fn lower(model: Model) -> Ir {
    let Model { trace_fields, item } = model;

    let fields = trace_fields
        .into_iter()
        .map(|expr| Field {
            message: format!("violation of trace_field `{}`", quote!(#expr)),
            expr,
        })
        .collect();

    Ir { fields, item }
}

pub struct Ir {
    pub fields: Vec<Field>,
    pub item: ItemFn,
}

pub struct Field {
    pub expr: syn::ExprAssign,
    pub message: String,
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use super::*;

    impl Model {
        fn stub() -> Self {
            Self {
                trace_fields: vec![],
                item: parse_quote!(
                    fn f() {}
                ),
            }
        }
    }

    #[test]
    fn produces_assertion_for_trace_field() {
        let mut model = Model::stub();
        model.trace_fields.push(parse_quote!(x = 2));

        let ir = lower(model);

        assert_eq!(1, ir.fields.len());

        let field = &ir.fields[0];
        let expected: syn::ExprAssign = parse_quote!(x=2);
        assert_eq!(expected, field.expr);
        assert_eq!("violation of trace_field `x = 2`", field.message);
    }
}
