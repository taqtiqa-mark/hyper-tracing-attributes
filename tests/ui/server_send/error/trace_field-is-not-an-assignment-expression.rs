use tracing_attributes_http::server_send;

#[server_send(trace, "Server")]
#[trace_field(b)]
fn f() {}

fn main() {}
