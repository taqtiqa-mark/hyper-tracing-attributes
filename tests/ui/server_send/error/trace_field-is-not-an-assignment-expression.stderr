warning: unused import: `abort_call_site`
 --> src/server_send/analyze.rs
  |
  | use proc_macro_error::{abort, abort_call_site};
  |                               ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `FromMeta`
 --> src/server_send/parsed.rs
  |
  | use darling::{FromMeta};
  |               ^^^^^^^^

warning: field is never read: `level`
  --> src/server_send/parsed.rs
   |
   |     level: Option<String>,
   |     ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: field is never read: `name`
  --> src/server_send/parsed.rs
   |
   |     name: Option<String>,
   |     ^^^^^^^^^^^^^^^^^^^^

warning: field is never read: `skip`
  --> src/server_send/parsed.rs
   |
   |     skip: Option<String>,
   |     ^^^^^^^^^^^^^^^^^^^^

warning: field is never read: `fields`
  --> src/server_send/parsed.rs
   |
   |     fields: Option<String>,
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: associated function is never used: `skip_params`
  --> src/server_send/parsed.rs
   |
   |     fn skip_params(&self) -> Vec<String> {
   |        ^^^^^^^^^^^

warning: associated function is never used: `fields_params`
  --> src/server_send/parsed.rs
   |
   |     fn fields_params(&self) -> Vec<String> {
   |        ^^^^^^^^^^^^^

trace, "Server"
2
Args { vars: [Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident { ident: "trace", span: #0 bytes(58..63) }, arguments: None }] } }), Lit(ExprLit { attrs: [], lit: Str(LitStr { token: "Server" }) })] }
error: this attribute takes from one to four arguments

         = help: use `#[server_send(level=trace, name="Name", skip="a, b")]`

 --> tests/ui/server_send/error/trace_field-is-not-an-assignment-expression.rs:3:15
  |
3 | #[server_send(trace,"Server")]
  |               ^^^^^

error[E0425]: cannot find function `f` in this scope
 --> tests/ui/server_send/error/trace_field-is-not-an-assignment-expression.rs:7:13
  |
7 | fn main() { f() }
  |             ^ not found in this scope
